frontend/src/test/scala/bloop/logging/RecordingLogger.scala

```diff
@@ -0,0 +1,26 @@
+package bloop.logging
+
+import scala.collection.mutable.Buffer
+
+class RecordingLogger extends AbstractLogger {
```

Comment:
Is it necessary to have a new logger for testing? I think there's a more elegant way to do this, and that is redirecting stdout in CLI.

Reply:
I use it to test `ProcessConfig`, which doesn't touch the CLI. There's something wrong with logging at the moment, and we can see it with the output being repeated several times in the tests.

When possible, I prefer to explicitly give a logger and inspect its content rather than setting an output stream somewhere and expecting to find some messages in it. When passing a `State` around, it should be enough to just change the underlying logger in the state.

Reply:
I agree that's a better approach, but it only works in some of the cases: not all bloop uses loggers to print stuff. If you fix the other stdout issue, then I agree this can stay here :smile:.
  
On the repeated messages in the loggers: I'll look into it.

Reply:
Oh, actually I haven't realized that it's better to set up the stdout, since the logger is redirected to the stdout anyway (don't you think this is better than breaking this assumption that things from the logger go to X and things to stdout go to Y?). I think that's simpler because everything in bloop relies on stdout being a nice place where you can redirect it and is accessible to the user. What do you think?

Reply:
I disagree. I don't think that a logger implementation should be required to eventually write stuff to stdout. Logging to a file, to the network, discarding all messages, etc. are all valid logger implementations in my eyes. Saying "Hey, this is stdout from now on" and expecting to see all output go there sounds much more fragile to me than just passing around the logger that you want to collect the output. It makes the path from message production to the logger much more obvious. It's probably also more friendly to parallel testing when we re-enable that (I'm not sure how much of L4J's global state we're mutating now).

In the future, I imagine that we should try to have one CLI invocation = one logger instance, set the right output stream for that logger, and put everything in the logger. I don't think that we should care about the actual output stream in use, we should just write to the logger and trust it to do the right thing.